---
author: "Marc Elosua-Bayes"
date: "9/23/2020"
output:
  html_document:
    toc: yes
    toc_float: yes
    number_sections: yes
    df_print: paged
params: 
    sample_id: "Default!"
    donor_id: "Default!"
title: "`r sprintf('%s QC', params$sample_id)`"
editor_options: 
  chunk_output_type: console
---

---
title: `r params$sample_id`
---

```{r setup, include=FALSE}
# knitr::opts_knit$set(root.dir = "../..")
knitr::opts_chunk$set(echo = TRUE, out.width = "100%", fig.align='center', 
                      message = FALSE, warning = FALSE, cache = FALSE)
options(width = 1200)
```

## Introduction

As we saw in the previous notebook, the samples have different sequencing depths. According to [Luecken MD et al.](https://www.embopress.org/doi/10.15252/msb.20188746), "if the distribution of QC covariates differ between samples, QC thresholds should be determined separately for each sample". This is the most reasonable approach, since empty droplets or lysed cells with high mitochondrial content will harbor different library size distribution across differentially sequenced samples, and hence need different QC thresholds.

The objective of this notebook is to filter out low-quality cells and genes of the cells coming from hashed samples.

## Load packages

```{r}
library(Seurat)
library(ggpubr)
library(cowplot)
library(dplyr)
library(ggplot2)
library(stringr)
library(readr)
library(plotly)
# library(tidyverse)
```

## Setting parameters
```{r}
source(here::here("misc/paths.R"))
```

```{r}
sample_id <- params$sample_id
# sample_id <- "qsbh45_pccb6n"
# THRESHOLDS
# min_lib_size <- 500
# min_n_genes <- 50
# max_pct_mt <- 50
# min_cells <- 8
min_lib_size <- qc_dict[[sample_id]][["min_lib_size"]]
max_lib_size <- qc_dict[[sample_id]][["max_lib_size"]]
min_n_genes <- qc_dict[[sample_id]][["min_n_genes"]]
max_pct_mt <- qc_dict[[sample_id]][["max_pct_mt"]]
min_cells <- qc_dict[[sample_id]][["min_cells"]]
scrublet_thresh <- qc_dict[[sample_id]][["scrublet_thresh"]]
```

## Load data

```{r}
cellranger_data <- glue::glue("{cell_ranger}/{sample_id}/{filtered_mtrx}")
count_mtrx <- Seurat::Read10X(here::here(cellranger_data))
scrublet <- readr::read_csv(
  file = here::here(glue::glue("{qc_28}/{ver}/scrublet_scores/scrublet_doublet_prediction-{sample_id}.csv"))) %>%
            tibble::column_to_rownames("barcodes")

# Create Seurat object
se_obj <- Seurat::CreateSeuratObject(counts = count_mtrx,
                                     project = sample_id,
                                     assay = "RNA",
                                     meta.data = scrublet)
```

Add mitochondrial and ribosomal percentage
```{r}
se_obj[["pct_mt"]] <- Seurat::PercentageFeatureSet(
  se_obj,
  pattern = "^mt-"
)

se_obj[["pct_ribo"]] <- Seurat::PercentageFeatureSet(
  se_obj,
  pattern = "^Rps"
)
```

## Filter out cells
We aim to detect and exclude empty droplets or lysed cells. Lysed cells have 3 hallmarks that we will exploit: 

1. low library size (total UMI)
2. low library complexity (number of detected genes)
3. high fraction of mitochondrial expression (cytosolic mRNA leaks out of the cell).

Let us start by visualize their univariate distribution.

### Library Size

```{r fig.wide=TRUE}
lib_size_hist1 <- se_obj@meta.data %>%
  ggplot(aes(nCount_RNA)) +
  geom_histogram(bins = 100) +
  labs(
    title = "Library Size Histogram",
    x = "Library Size (log10(totalUMI))",
    y = "Number of Cells"
    ) +
  scale_x_log10() +
  geom_vline(xintercept = 1500) +
  ggpubr::theme_pubr()
lib_size_hist1

lib_size_hist2 <- se_obj@meta.data %>%
  ggplot(aes(nCount_RNA)) +
    geom_histogram(bins = 100) +
    # scale_x_continuous(limits = c(0, 4000)) +
    labs(x = "Library Size (total UMI)", y = "Number of Cells") +
    ggpubr::theme_pubr()
lib_size_hist2

cowplot::plot_grid(plotlist = list(lib_size_hist1, lib_size_hist2),
                   ncol = 2)
```

### Number of detected genes

```{r fig.wide=TRUE}
n_genes_hist1 <- se_obj@meta.data %>%
  ggplot(aes(nFeature_RNA)) +
    geom_histogram(bins = 100) +
    labs(x = "Number of Detected Genes (log(nFeat))", y = "Number of Cells") +
    scale_x_log10() +
    ggpubr::theme_pubr()
n_genes_hist1

n_genes_hist2 <- se_obj@meta.data %>%
  ggplot(aes(nFeature_RNA)) +
    geom_histogram(bins = 100) +
    # scale_x_continuous(limits = c(0, 2000)) +
    labs(x = "Number of Detected Genes", y = "Number of Cells") +
    ggpubr::theme_pubr()
n_genes_hist2

cowplot::plot_grid(plotlist = list(n_genes_hist1, n_genes_hist2),
                   ncol = 2)
```

We can start to glimpse that there is a bimodal distribution within this data.
We're gonna need to keep an eye out while cleaning this dataset!

### Fraction of mitochondrial expression

```{r}
pct_mt_hist <- se_obj@meta.data %>%
  ggplot(aes(pct_mt)) +
    geom_histogram(bins = 100) +
    scale_x_continuous(limits = c(0, 100)) +
    labs(x = "% Mitochondrial Expression", y = "Number of Cells") +
    ggpubr::theme_pubr()
pct_mt_hist
```

### Fraction of ribosomal expression

```{r}
pct_ribo_hist <- se_obj@meta.data %>%
  ggplot(aes(pct_ribo)) +
    geom_histogram(bins = 100) +
    scale_x_continuous(limits = c(0, 100)) +
    labs(x = "% Ribosomal Expression", y = "Number of Cells") +
    ggpubr::theme_pubr()
pct_ribo_hist
```

### Scrublet score
We can take a look at the [scrublet vignette](https://github.com/AllonKleinLab/scrublet/blob/master/examples/scrublet_basics.ipynb).

When looking at the scrublet score yhe simulated doublet histogram is typically bimodal. The left mode corresponds to "embedded" doublets generated by two cells with similar gene expression. The right mode corresponds to "neotypic" doublets, which are generated by cells with distinct gene expression (e.g., different cell types) and are expected to introduce more artifacts in downstream analyses. Scrublet can only detect neotypic doublets.

To call doublets vs. singlets, we must set a threshold doublet score, ideally at the minimum between the two modes of the simulated doublet histogram. scrub_doublets() attempts to identify this point automatically and has done a good job in this example. However, if automatic threshold detection doesn't work well, you can adjust the threshold with the call_doublets() function.

```{r}
scrublet_hist <- se_obj@meta.data %>%
  ggplot(aes(scrublet_doublet_scores)) +
    geom_histogram(bins = 100) +
    scale_x_continuous(limits = c(0, 1)) +
    labs(x = "Scrublet score", y = "Number of Cells") +
    ggpubr::theme_pubr()

scrublet_hist
```

### Joint QC metrics
It is important to assess how these variables covary, since metabolically active cells might also have a high mitochondrial expression:

We check where high mitochondrial expressing genes are found compared to Library size and number of detected genes.
```{r}
# number of detected genes VS library size
n_genes_vs_lib_size <- ggplot(se_obj@meta.data,
       aes(x = nCount_RNA,
           y = nFeature_RNA,
           color = pct_mt)) +
  geom_point() +
  ggpubr::stat_cor() +
  ggpubr::theme_pubr() +
  scale_color_viridis_c() +
  labs(
    x = "Library Size (total UMI)",
    y = "Number of Detected Genes",
    color = "Mitochondrial %") +
  geom_vline(xintercept = min_lib_size, linetype = "dashed", color = "red") +
  geom_vline(xintercept = max_lib_size, linetype = "dashed", color = "red") +
  geom_hline(yintercept = min_n_genes, linetype = "dashed", color = "red")

plotly::ggplotly(n_genes_vs_lib_size)
```

We want to see where the doublets are found, if we see high scrublet scores at the top right it gives a clue that they are doublets since they have a high score and they have a large library size and number of detected genes.
```{r}
# number of library size VS number of genes
n_genes_vs_lib_size_scrublet <- ggplot(se_obj@meta.data,
       aes(x = nCount_RNA,
           y = nFeature_RNA,
           color = scrublet_doublet_scores)) +
  geom_point() +
  ggpubr::stat_cor() +
  ggpubr::theme_pubr() +
  scale_color_viridis_c() +
  labs(
    x = "Library Size (total UMI)",
    y = "Number of Detected Genes",
    color = "Scrublet score")+
  geom_vline(xintercept = min_lib_size, linetype = "dashed", color = "red") +
  geom_vline(xintercept = max_lib_size, linetype = "dashed", color = "red") +
  geom_hline(yintercept = min_n_genes, linetype = "dashed", color = "red")

plotly::ggplotly(n_genes_vs_lib_size_scrublet)
```

```{r}
# % mitochondrial expression VS library size
pct_mt_vs_lib_size_scrublet <- ggplot(se_obj@meta.data,
       aes(x = nCount_RNA,
           y = pct_mt,
           color = scrublet_doublet_scores)) +
  geom_point() +
  ggpubr::stat_cor() +
  ggpubr::theme_pubr() +
  scale_color_viridis_c() +
  labs(
    x = "Library Size (total UMI)",
    y = "Mitochondrial %",
    color = "Scrublet score") +
  geom_vline(xintercept = min_lib_size, linetype = "dashed", color = "red") +
  geom_vline(xintercept = max_lib_size, linetype = "dashed", color = "red") +
  geom_hline(yintercept = max_pct_mt, linetype = "dashed", color = "red")

plotly::ggplotly(pct_mt_vs_lib_size_scrublet)
```

```{r}
# Plot thresholds in the histograms
plotly::ggplotly(lib_size_hist1 +
  geom_vline(xintercept = min_lib_size,
             linetype = "dashed",
             color = "red") +
  geom_vline(xintercept = max_lib_size,
             linetype = "dashed",
             color = "red") +
  ggpubr::theme_pubr())

plotly::ggplotly(n_genes_hist1 +
  geom_vline(xintercept = min_n_genes,
             linetype = "dashed",
             color = "red") +
  ggpubr::theme_pubr())

plotly::ggplotly(pct_mt_hist +
  geom_vline(xintercept = max_pct_mt,
             linetype = "dashed",
             color = "red") +
  ggpubr::theme_pubr())

plotly::ggplotly(scrublet_hist +
  geom_vline(xintercept = scrublet_thresh,
             linetype = "dashed",
             color = "red") +
  ggpubr::theme_pubr())
```


```{r}
metadata_before_qc <- se_obj@meta.data
is_low_quality <- 
  se_obj$nCount_RNA < min_lib_size |
  se_obj$nCount_RNA > max_lib_size |
  se_obj$nFeature_RNA < min_n_genes |
  se_obj$pct_mt > max_pct_mt |
  se_obj$scrublet_doublet_scores > scrublet_thresh
table(is_low_quality)

se_obj$keep_cells <- !is_low_quality
```
output:
  html_document:
    toc: yes
    toc_float: yes
    number_sections: yes
    df_print: paged

Visualize on the UMAP where these cells are embedded
```{r}
se_obj <- Seurat::NormalizeData(se_obj,
                                   verbose = FALSE) %>%
  Seurat::FindVariableFeatures(., verbose = FALSE) %>%
  Seurat::ScaleData(., verbose = FALSE) %>%
  Seurat::RunPCA(., verbose = FALSE) %>% 
  Seurat::RunUMAP(., reduction = "pca",
                  dims = 1:30)

Seurat::DimPlot(se_obj,
                group.by = "keep_cells",
                cols = c("red", "black"))
```

### Subset
```{r}
Idents(se_obj) <- "keep_cells"
se_obj <- subset(se_obj, idents = TRUE)
metadata_after_qc <- se_obj@meta.data
```


### QC summary table
I need a stronger traceback of the number of cells I exclude at each step, including:
* Num of cells excluded in demultiplexing
* Num of cells excluded in doublet detection
* Num of cells excluded filtering poor quality cells
```{r}
qc_before <- metadata_before_qc %>%
  # group_by(gem_id) %>% 
  summarise(
    num_cells_before_qc = n(),
    average_library_size = mean(nCount_RNA),
    average_num_detected_genes = mean(nFeature_RNA),
    average_mitochondrial_fraction = mean(pct_mt)
)
qc_before

qc_after <- metadata_after_qc %>%
  # group_by(gem_id) %>%
  summarise(
    num_cells_after_qc = n(),
    average_library_size = mean(nCount_RNA),
    average_num_detected_genes = mean(nFeature_RNA),
    average_mitochondrial_fraction = mean(pct_mt)
  )

qc_after
# qc_table <- left_join(qc_before, qc_after, by = "gem_id")
# DT::datatable(qc_table)
```

## Filter out genes

According to [Luecken MD et al.](https://www.embopress.org/doi/10.15252/msb.20188746): "A guideline to setting this threshold is to use the minimum cell cluster size that is of interest and leaving some leeway for dropout effects. For example, filtering out genes expressed in fewer than 20 cells may make it difficult to detect cell clusters with fewer than 20 cells. For datasets with high dropout rates, this threshold may also complicate the detection of larger clusters. The choice of threshold should scale with the number of cells in the dataset and the intended downstream analysis."

Since we want to detect rare cell types, we will be very permissive and set retain genes that are expressed in at least `r min_cells` cells.

### Set threshold

```{r fig.wide=TRUE}
n_cells <- Matrix::rowSums(se_obj[["RNA"]]@counts > 0)
gene_qc <- n_cells %>% 
  as.data.frame() %>% 
  ggplot(aes(n_cells)) + 
    geom_histogram(bins = 100, alpha = 0.75) +
    scale_x_log10("Number of cells") +
    theme_bw() 

plotly::ggplotly(gene_qc +
  geom_vline(xintercept = min_cells, linetype = "dashed", color = "red"))
```

### Plot highest genes expressed throughout

```{r fig.height=10}
top_50_genes <- sort(n_cells, decreasing = TRUE)[1:50]
top_50_genes_df <- data.frame(
  gene = names(top_50_genes),
  n_cells = top_50_genes
)
top_50_genes_df %>%
  ggplot(aes(forcats::fct_reorder(gene, n_cells), n_cells)) +
    geom_point() +
    labs(x = "", y = "Number of expressing cells") +
    coord_flip() +
  ggpubr::theme_pubr()
```

### Filter

```{r}
kept_genes <- rownames(se_obj)[n_cells > min_cells]
table(n_cells > min_cells)
se_obj <- subset(se_obj, features = kept_genes)
se_obj
```

## Dimensionality reduction + clustering
Scale and Normalize the data using SCTransform
```{r fig.height=15, fig.width=15}
se_obj <- Seurat::SCTransform(object = se_obj,
                              assay = "RNA",
                              verbose = FALSE)

se_obj <- Seurat::RunPCA(se_obj)
print(se_obj[["pca"]], dims = 1:10, nfeatures = 5)
Seurat::VizDimLoadings(se_obj, dims = 1:10, reduction = "pca")
```

Determining the dimensionality of the dataset
```{r}
Seurat::ElbowPlot(object = se_obj, ndims = 50)
```

We are going to use 30 PCs
```{r, fig.height=7, fig.width=10}
se_obj <- Seurat::FindNeighbors(se_obj, dims = 1:30) %>%
  Seurat::FindClusters(., resolution = c(0.01, 0.05, 0.1, 0.3, 0.8))
se_obj <- Seurat::RunUMAP(se_obj, dims = 1:30)

dim_plt <- Seurat::DimPlot(object = se_obj,
                          group.by = c("SCT_snn_res.0.01",
                                       "SCT_snn_res.0.05",
                                       "SCT_snn_res.0.1",
                                       "SCT_snn_res.0.3",
                                       "SCT_snn_res.0.8"),
                          label = TRUE,
                          ncol = 3)
dim_plt
```

Look at how number of UMIs, number of genes, mitochondrial % and scrublet score on the UMAP embedding
```{r fig.height=8, fig.width=8}
feat_plt <- Seurat::FeaturePlot(object = se_obj,
                                features = c("nCount_RNA",
                                             "nFeature_RNA",
                                             "pct_mt",
                                             "scrublet_doublet_scores"),
                                ncol = 2)
feat_plt
```

### Cell cycle scoring
```{r fig.height=8, fig.width=8}
# A list of cell cycle markers, from Tirosh et al, 2015, is loaded with Seurat.  We can
# segregate this list into markers of G2/M phase and markers of S phase
s.genes <- cc.genes$s.genes
g2m.genes <- cc.genes$g2m.genes

se_obj <- Seurat::RunPCA(se_obj,
                         features = VariableFeatures(se_obj),
                         # ndims.print = 6:10,
                         nfeatures.print = 10)

se_obj <- Seurat::CellCycleScoring(se_obj,
                                   s.features = s.genes,
                                   g2m.features = g2m.genes,
                                   set.ident = TRUE)

cc_plt <- Seurat::DimPlot(object = se_obj,
                          group.by = "Phase")

feat_plt[[2]] + feat_plt[[3]] + dim_plt[[1]] + cc_plt
```

## Save filtered Seurat object

```{r}
saveRDS(object = se_obj,
        file = here::here(glue::glue("{qc_28}/{robj_dir}/{sample_id}_qc.rds")))
```

## Session Information

```{r}
sessionInfo()
```