---
title: "01-QC metrics summary"
author: "Marc Elosua-Bayes"
date: "9/28/2020"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE, out.width = "100%", fig.align='center', 
                      message = FALSE, warning = FALSE)
options(width = 1200)
```

## Introduction

EPID_27 is a project aimed to
The objective of this notebook is to perform a basic quality control (QC) analysis of the mapping performed with [spaceranger 1.1.0](https://support.10xgenomics.com/spatial-gene-expression/software/pipelines/latest/using/count).

## Load packages

```{r}
library(ggpubr)
library(ggrepel)
library(DT)
library(dplyr)
library(readr)
library(ggplot2)
library(stringr)
```

## Parameters
Loading necessary paths and parameters
```{r}
source(here::here("misc/paths.R"))
plt_dir <- sprintf("%s/%s", qc_27, plt_dir)
robj_dir <- sprintf("%s/%s", qc_27, robj_dir)

dir.create(path = here::here(plt_dir),
           showWarnings = FALSE,
           recursive = TRUE)

dir.create(path = here::here(robj_dir),
           showWarnings = FALSE,
           recursive = TRUE)
```


## Load data
We have 4 different datasets, 1 for each capture area.
```{r}
# Load ID matcher
id_df <- readr::read_csv(here::here("spatial_analysis/01-spaceranger/data/sample_id.txt"))

qcmetric_df <- lapply(sp_id_ls, function(id) {
  tmp_df <- readr::read_csv(file = here::here(sprintf("%s/%s/outs/metrics_summary.csv",
                                           spaceranger, id)))
  tmp_df <- tmp_df %>%
    dplyr::mutate(sample_id = id_df[id_df$gem_id == id, ] %>% pull(donor_id))
  
  return(tmp_df)
}) %>% dplyr::bind_rows()

colnames(qcmetric_df) <- stringr::str_replace_all(string = colnames(qcmetric_df),
                                                  pattern = " ",
                                                  replacement = "_")

DT::datatable(qcmetric_df)
```

# Sequencing

Our first objective is to quantify the quality of our sequenced libraries prior to mapping. We will leverage the "Q30" variables in our dataset. According to 10X, these report the fraction of bases with a Q-score of at least 30 for different sequences: cell barcode, RNA reads, Unique Molecular Identifiers (UMI). Q-score is calculated as follows:

$$Q = -10\log10(p)$$
Where p is the probability of the base being wrongly called. Thus, bases with a high Q are highly reliable.

```{r fig.height=18}
# SEQUENCING QC
cols <- c("#cb3b25", "#b930df")

## Q30
q30_vars <- stringr::str_subset(colnames(qcmetric_df), "Q30")

q30_gg <- purrr::map(q30_vars, function(var) {
  
  tmp_df <- qcmetric_df %>%
    dplyr::mutate(
      var_int = stringr::str_remove(string = qcmetric_df %>% dplyr::pull(all_of(var)),
                                pattern = "%"),
      var_int = as.numeric(var_int)
    )

  ggplot(tmp_df,
         aes(x = "1",
             y = var_int * 100)) +
    # geom_boxplot(fill = NA,
    #              outlier.shape = NA) +
    geom_jitter(aes(color = sample_id),
                width = 0.1,
                height = 0) +
    ylim(c(75, 100)) +
    labs(x = "",
         y = stringr::str_c(stringr::str_replace_all(var, "_", " "), " (%)")) +
    ggpubr::theme_pubr() +
    theme(axis.title = element_text(size = 13),
          axis.text = element_text(size = 13,
                                   color = "black"))
})

# q30_gg
q30_gg_arr <- ggarrange(plotlist = q30_gg, nrow = 3, ncol = 1)
q30_gg_arr
```

# Mapping

Secondly, we will assess the quality of cellranger's mapping by comparing the percentage of reads mapping to the genome, intergenic regions, intronic and exonic regions across libraries. Reads mapping to intergenic regions suggest contamination of ambient DNA, while reads mapping to intronic regions may come from pre-mRNAs or mature splice isoforms that retain the introns:

```{r fig.wide=TRUE, fig.height=12}
# Reads mapped to genome and transcriptome
mapping_qc_vars <- c(
  "Reads_Mapped_to_Genome_(mm10)",
  "Reads_Mapped_Confidently_to_Intergenic_Regions_(mm10)",
  "Reads_Mapped_Confidently_to_Intronic_Regions_(mm10)",
  "Reads_Mapped_Confidently_to_Exonic_Regions_(mm10)",
  "Reads_Mapped_to_Genome_(GRCh38)",
  "Reads_Mapped_Confidently_to_Intergenic_Regions_(GRCh38)",
  "Reads_Mapped_Confidently_to_Intronic_Regions_(GRCh38)",
  "Reads_Mapped_Confidently_to_Exonic_Regions_(GRCh38)"
)

mapping_qc_gg <- purrr::map(mapping_qc_vars , function(var) {
  qcmetric_df <- qcmetric_df %>%
  dplyr::mutate(
    var_int = stringr::str_remove(string = qcmetric_df %>%
                                    dplyr::pull(all_of(var)),
                              pattern = "%"),
    var_int = as.numeric(var_int)
  )

  ggplot(qcmetric_df,
         aes_string("sample_id",
                    "var_int",
                    color = "sample_id")) +
    geom_point() +
    labs(x = "", y = stringr::str_c(stringr::str_replace_all(var, "_", " "), " (%)")) +
    theme_bw() +
    theme(legend.position = "none",
          axis.title = element_text(size = 13),
          axis.text = element_text(size = 13, color = "black"),
          axis.text.x = element_text(angle = 45, vjust = 1, hjust = 1))
})

# mapping_qc_gg
mapping_qc_arr <- ggarrange(plotlist = mapping_qc_gg, nrow = 2, ncol = 2)
mapping_qc_arr
```

# Sequencing Saturation

Thirdly, we will plot the number of detected genes per library as a function of the total reads sequenced. We know that this function reaches a plateau, whereby more sequenced reads does not result in more detected genes. In those scenarios, we say that we have sequenced until saturation:

```{r fig.width = 8, fig.height = 6}
selected_qc <- qcmetric_df[qcmetric_df$`Total_Genes_Detected_(GRCh38)` + 
                             qcmetric_df$`Total_Genes_Detected_(mm10)` < 20000, ]

num_genes_vs_total_reads <- qcmetric_df %>%
  dplyr::mutate(Total_Genes_Detected = `Total_Genes_Detected_(GRCh38)` + `Total_Genes_Detected_(mm10)`) %>%
  ggplot(aes(Number_of_Reads, Total_Genes_Detected, color = sample_id)) +
    geom_point() +
    # geom_text_repel(data = selected_qc, aes(label = sample_id), color = "black") +
    # scale_color_manual("", values = cols) +
    labs(x = "Number of Reads", y = "Total Genes Detected", color = "") +
    theme_classic() +
    theme(axis.title = element_text(size = 13, color = "black"),
          axis.text = element_text(size = 12, color = "black"),
          legend.text = element_text(size = 12, color = "black"))
num_genes_vs_total_reads
```

# Session Info

```{r}
sessionInfo()
```